plugin-java/che-plugin-java-ext-jdt/che-jdt-ext-machine/src/main/java/org/eclipse/che/jdt/refactoring/session/RefactoringSession.java

```diff
@@ -112,23 +127,109 @@ public Change getChange() {
         return change;
     }
 
-    public RefactoringStatus apply() {
+    /**@return instance of {@link org.eclipse.che.ide.ext.java.shared.dto.refactoring.RefactoringStatus}.
+     * That describes status of the refactoring operation. */
+    public org.eclipse.che.ide.ext.java.shared.dto.refactoring.RefactoringStatus apply() {
         PerformChangeOperation operation = new PerformChangeOperation(change);
         FinishResult result = internalPerformFinish(operation);
         if (result.isException()) {
-            return RefactoringStatus.createErrorStatus("Refactoring failed with Exception.");
+            return DtoConverter.toRefactoringStatusDto(RefactoringStatus.createErrorStatus("Refactoring failed with Exception."));
+        }
+
+        CompositeChange operationChange = (CompositeChange)operation.getUndoChange();
+        Change[] changes = operationChange.getChildren();
+
+        RefactoringStatus validationStatus = operation.getValidationStatus();
+        if (validationStatus != null) {
+            List<ChangeInfo> changesInfo = new ArrayList<>();
+
+            prepareChangesInfo(changes, changesInfo);
+
+            org.eclipse.che.ide.ext.java.shared.dto.refactoring.RefactoringStatus status =
+                    DtoConverter.toRefactoringStatusDto(validationStatus);
+            status.setChanges(changesInfo);
+
+            return status;
+        }
+
+        return DtoConverter.toRefactoringStatusDto(new RefactoringStatus());
+    }
+
+    private void prepareChangesInfo(Change[] changes, List<ChangeInfo> changesInfo) {
+        for (Change ch : changes) {
+            if (ch instanceof DynamicValidationStateChange) {
+                prepareChangesInfo(((DynamicValidationStateChange)ch).getChildren(), changesInfo);
+            } else {
+                ChangeInfo changeInfo = DtoFactory.newDto(ChangeInfo.class);
+                String refactoringName = ch.getName();
+                if (ch instanceof UndoTextFileChange) {
+                    changeInfo.setName(ChangeInfo.ChangeName.UPDATE);
+                    changeInfo.setPath(((CompilationUnit)ch.getModifiedElement()).getPath().toString());
+                }
+                if (refactoringName.startsWith("Rename")) {
+                    if (ch instanceof RenameCompilationUnitChange) {
```

Comment:
It can, but these classes are from eclipse.jdt.* package and I don't want to make some big changes into these files. I didn't use 'if else' because we have more then 2 types of the Rename refactoring but we don't yet supporting it (for example rename source folder).